cpp/include/raft/neighbors/cagra.cuh

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "detail/cagra/cagra_build.cuh"
+#include "detail/cagra/cagra_search.cuh"
+#include "detail/cagra/graph_core.cuh"
+
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/device_resources.hpp>
+#include <raft/core/host_device_accessor.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/neighbors/cagra_types.hpp>
+#include <rmm/cuda_stream_view.hpp>
+
+namespace raft::neighbors::experimental::cagra {
+
+/**
+ * @defgroup cagra CUDA ANN Graph-based nearest neighbor search
+ * @{
+ */
+
+/**
+ * @brief Build a kNN graph.
+ *
+ * The kNN graph is the first building block for CAGRA index.
+ * This function uses the IVF-PQ method to build a kNN graph.
+ *
+ * See [cagra::build](#cagra::build) for alternative method.
+ *
+ * The following distance metrics are supported:
+ * - L2Expanded
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   using namespace raft::neighbors;
+ *   // use default index parameters
+ *   ivf_pq::index_params build_params;
+ *   ivf_pq::search_params search_params
+ *   auto knn_graph      = raft::make_host_matrix<IdxT, IdxT>(dataset.extent(0), 128);
+ *   // create knn graph
+ *   cagra::build_knn_graph(res, dataset, knn_graph.view(), 2, build_params, search_params);
+ *   auto pruned_gaph      = raft::make_host_matrix<IdxT, IdxT>(dataset.extent(0), 64);
+ *   cagra::prune(res, dataset, knn_graph.view(), pruned_graph.view());
+ *   // Construct an index from dataset and pruned knn_graph
+ *   auto index = cagra::index<T, IdxT>(res, build_params.metric(), dataset, pruned_graph.view());
+ * @endcode
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices in the source dataset
+ *
+ * @param[in] res raft resources
+ * @param[in] dataset a matrix view (host or device) to a row-major matrix [n_rows, dim]
+ * @param[out] knn_graph a host matrix view to store the output knn graph
+ * @param[in] refine_rate refinement rate for ivf-pq search
+ * @param[in] build_params (optional) ivf_pq index building parameters for knn graph
+ * @param[in] search_params (optional) ivf_pq search parameters
+ */
+template <typename DataT, typename IdxT, typename accessor>
+void build_knn_graph(raft::device_resources const& res,
+                     mdspan<const DataT, matrix_extent<IdxT>, row_major, accessor> dataset,
+                     raft::host_matrix_view<IdxT, IdxT, row_major> knn_graph,
```

Comment:
Is this a sparse or dense data structure?

Reply:
A dense matrix that stores the neighbor indices, same number of neighbors for all points.